### ISSUE TITLE
Duplicate notebook file after renaming while dirty

### ISSUE THREAD

**Issue by chrmarti** [microsoft/vscode#101370]
Testing #101295

- Auto-save turned off.
- Open notebook and change to dirty.
- Rename notebook file in Explorer.
- Save notebook.
- New copy appears under old name.

**Comment by rebornix** [comment]
Good catch and I wonder if it's a common issue of custom editors which use non-file scheme resource uri for working copy cc @mjbvz 

What's happening under the hood is 

* Rename the file
* `WorkingCopyFileCopyService` tries to do soft revert if it find a dirty working copy for the file https://github.com/microsoft/vscode/blob/8e66829775a449c00908fc82698781911c724f8f/src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts#L296-L299
* For text file
  * Soft revert just makes it not dirty
  * Then editor service will close the input and open a new one
* For notebook (and potentially custom binary editor)
  * It can't find the working copy as `workingCopy.resource !== fileResource`
  * It renames the existing editor input but does not open the new one in the editor pane

So if you have a dirty notebook `1.github-issues`, rename it to `2.github-issues`, the file in the editor pane is still `1.github-issues`.

**Comment by eamodio** [comment]
I see the same thing, with a normal text custom editor (CustomTextEditorProvider) -- I'm using the GitLens Rebase editor to test. Although for me, it doesn't matter if the document is dirty or not.

**Comment by rebornix** [comment]
We need an unified solution for this.

**Comment by bpasero** [comment]
> We need an unified solution for this.

This will be something the new file based working copies can help with. We have complicated code for text file models that tracks rename operations for dirty models to restore them after the operation has finished:

https://github.com/microsoft/vscode/blob/4c9d23c8f57748e8cef193644dc9165900d28efa/src/vs/workbench/services/textfile/common/textFileEditorModelManager.ts#L92-L95

I plan to bring this code over into the file working copy manager that @jrieken is adopting.

**Comment by jrieken** [comment]
> We need an unified solution for this.

I think notebooks are special because we know the model structure, just not how to get from data to model. Notebooks using the upcoming serializer-API should be all covered as @bpasero mentioned above. Then there is notebook using the complex data provider. I see the following dimensions:

* Is the notebook after rename still a notebook? E.g do we have any content provider for the new name?
* Is the notebook dirty?

When dealing with saved (not dirty) notebooks everything is easy, but there are challenges when notebooks are dirty:

1. notebook is dirty but will stay a notebook
1. notebook is dirty and will not be opened as notebook

For case 1 we can try to transfer the existing notebook model onto the target model but for case 2 we don't know what to do. Best would be to save _before_ renaming but AFAIK the rename already happened when `EditorInput#rename` is called (maybe @bpasero knows more) and then we are going to loose the data

**Comment by bpasero** [comment]
I am open to add more support from the workbench to make this easier, today the responsibility of what happens for a rename is split between editor service and working copy file service. In order of what happens:
* user renames a file via explorer or from a bulk edit (this is required for the working copy file service to kick in)
* there is a working copy event to handle the rename before it actually happens and this is the code I mentioned in https://github.com/microsoft/vscode/issues/101370#issuecomment-803280631 to remember the file working copies to restore (I would think a similar listening mechanism could be implemented for complex notebooks to do something before the rename happens)
* the rename happens on the level of the file system
* the editor services calls `EditorInput#rename` from [`handleMovedFile`](https://github.com/microsoft/vscode/blob/c72d7612b412031d8b6003acac58a0554ea1ca09/src/vs/workbench/services/editor/browser/editorService.ts#L226-L226)
* there is another working copy event for when the operation has finished and this is the point where working copy files restore with the new file path and the previous contents if they were dirty

**tl;dr;** I feel like complex notebooks would need to participate in the before and after of the working copy file operation. I guess the challenge is that for complex notebooks you have no clue what the files are.

**Comment by lramos15** [comment]
Seems like this has been fixed for notebooks. Marking as important though as renaming a dirty custom editor causes data loss. /cc @mjbvz

**Comment by jrieken** [comment]
> Seems like this has been fixed for notebooks. 

This is only partially fixed for notebooks - it works for those that use the serializer API (which we strongly recommend) but it doesn't work for notebooks using the old content provider API

**Comment by lramos15** [comment]
Removing important as the data loss case for custom editors is now fixed

**Comment by lramos15** [comment]
What's the status of this issue now? @rebornix do you know? It sounds like once all notebooks transition to the new model this isn't an issue anymore. Is the old content provider API being deprecated?

**Comment by mjbvz** [comment]
I believe this is now working as expected? I just tested with the api notebooks and a jupyter notebook

Closing for verification
